ContactDetailsApi/V2/Gateways/ContactDetailsDynamoDbGateway.cs

```diff
@@ -101,5 +114,344 @@ public async Task<ContactDetails> CreateContact(ContactDetailsEntity contactDeta
+                        .Select(x => new PersonContactDetails
+                        {
+                            ContactType = x.ContactInformation.ContactType.ToString(),
+                            SubType = x.ContactInformation.SubType.ToString(),
+                            Value = x.ContactInformation.Value.ToString()
+                        })
+                        .ToList();
+
+
+                    var personContact = new Person
+                    {
+                        PersonTenureType = householdMember.PersonTenureType.ToString(),
+                        IsResponsible = householdMember.IsResponsible,
+                        FirstName = person.FirstName,
+                        LastName = person.Surname,
+                        Title = person.Title.ToString(),
+                        PersonContactDetails = personContactDetails
+                    };
+
+                    personContacts.Add(personContact);
+                }
+
+                var contactByUprn = new ContactByUprn
+                {
+                    Uprn = asset.Uprn,
+                    TenureId = tenure.Id,
+                    Contacts = personContacts
+                };
+
+                contacts.Add(contactByUprn);
+            }
+            return contacts;
+        }
+
+        [LogCall]
+        public async Task<List<ContactByUprn>> FetchAllContactDetailsByUprnUseCase()
+        {
+            // 1. Scan all assets
+            var assets = await FetchAllAssets();
+
+            // remove assets without a uprn
+            var filteredAssets = assets
+                .Where(x => !string.IsNullOrWhiteSpace(x.Uprn))
+                .ToList();
+
+            // 2. Fetch all contact details
+            var contactDetails = await FetchAllContactDetails();
+
+
+            var contactDetailsGroupedByTargetId = contactDetails
+                .GroupBy(c => c.TargetId)
+                .ToDictionary(group => group.Key, group => group.ToList());
+
+            // 2. Fetch tenure records
+            var filterTenures = assets.Where(x => !string.IsNullOrWhiteSpace(x.TenureId.ToString())).ToList();
+            var tenureIds = filterTenures.Select(x => x.TenureId).Distinct().ToList();
+            var tenures = await FetchTenures(tenureIds);
+            var tenuresByTenureId = tenures.ToDictionary(x => x.Id, x => x);
+
+            // 3. For each household member, fetch contact details, and person recond
+            var personIds = tenures.SelectMany(tenure => tenure.HouseholdMembers.Select(x => x.Id)).Distinct().ToList();
+            var persons = await FetchPersons(personIds);
+            var personById = persons.ToDictionary(x => x.Id, x => x);
+
+            // 4. consolidate the data
+            var contacts = GetContactByUprnForEachAsset(assets, tenuresByTenureId, personById, contactDetailsGroupedByTargetId);
+
+            return contacts;
+        }
```

Comment:
Could we move this logic to the usecase, as its not interacting with the database directly except via each of the public methods it calls?

Reply:
will do this in another pr